### DISCUSSION TITLE
Should amp be added as an eviction algorithm or a prefetch algorithm?

### DISCUSSION THREAD

**Discussion by zztaki** [1a1a11a/libCacheSim#22] [Ideas]
Hi, I am confused that amp should be a prefetch-aware replacement algorithm or a prefetch algorithm.

In [the original paper](https://www.usenix.org/conference/fast-07/amp-adaptive-multi-stream-prefetching-shared-cache), the authors not only do "Theoretical Analysis: Optimality Criteria" under LRU cache replacement algorithm, but also add a trick, objects are prefetched into the cache will not be promoted when they are accessed in the first time. Both of these methods make amp strongly coupled with the cache replacement algorithm.

I'm confused about whether unbinding the amp from the lru and removing this trick would cause a semantic error.

**Comment by 1a1a11a** [comment]
There are a few solutions. For example, we can 
1. implement AMP as a prefetch algorithm, which marks an object as not promotable when prefetched, and changes to promotable upon second request. This can be used with any eviction algorithm. 
2. add an AMP eviction algorithm in `cachesim` command line option which uses AMP and LRU. 

Does this make sense?

**Reply by zztaki** [reply]
Yes, I think both options are very attractive.

Meanwhile, I think option 1 is better because it is a superset of option 2. As for whether there will be a semantic error in combination with non-lru algorithms, this should be left to the user who uses the combination to judge in his context.

But does it mean that I need to modify the **{algo}_find** function of all eviction algorithms and add a new member to object structure?

**Reply by 1a1a11a** [reply]
Do you think we can leave {algo}_find as it is, and modify the `cache_get_base` function? 
we can add if to this line so that if the prefetcher is AMP, then we find with `update_data` false first, and then `find` again if we need to promote the object. 
https://github.com/1a1a11a/libCacheSim/blob/5fa68ef6902350aa9734398862bec7912d59f5e3/libCacheSim/cache/cache.c#L209

**Reply by zztaki** [reply]
Yes, you are right! In this case, we only need to add a structure member, and this structure member is only visible to amp temporarily, and other algorithms do not need to maintain it.

But this seems to put more and more burden on the common interface in the `cache.c`. It is really a bit upset to add the prefetch algorithm. :(

**Reply by 1a1a11a** [reply]
You are right. You may find a better idea while implementing this, I am happy to discuss other approaches. :)
